### PR TITLE
maybe perf tuning | strict ifs

### PR DESCRIPTION
**Pull request by lifeart**
@pzuraq is it make sence?

// https://github.com/glimmerjs/glimmer-vm/pull/964

### CODE REVIEW

**Comment by lifeart** [comment]
possible rule - https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md

**Comment by krisselden** [comment]
in general for hotspots the explicit checks are better than falsy coercions, especially if you have several types that falsy coerce and one of them happens occasionally.

File: packages/@glimmer/reference/lib/reference.ts [hunk 1]
```diff
@@ -59,22 +59,22 @@ export class ReferenceCache<T> implements Tagged {
...
-    if (!this.initialized) {
+    if (this.initialized === false) {
       return this.initialize();
     }
 
     let { reference, lastRevision } = this;
     let tag = reference.tag;
 
-    if (validate(tag, lastRevision as number)) return NOT_MODIFIED;
+    if (validate(tag, lastRevision as number) === true) return NOT_MODIFIED;
```
Comment (krisselden) [inline packages/@glimmer/reference/lib/reference.ts]: does validate not always return a boolean? if validate always returns a boolean then there is no benefit from this

File: packages/@glimmer/runtime/lib/vm/stack.ts [hunk 2]
```diff
@@ -40,8 +40,8 @@ export class InnerStack {
   }
 
   write(pos: number, value: unknown): void {
-    if (isImmediate(value)) {
-      this.writeRaw(pos, encodeImmediate(value));
+    if (isImmediate(value) === true) {
```
Comment (krisselden) [inline packages/@glimmer/runtime/lib/vm/stack.ts]: if isImmediate always returns a boolean then there is no coercion you are avoiding by doing this.

**Comment by pzuraq** [comment]
@lifeart can you add that eslint rule in this PR? It should allow the cases that Kris pointed out, where we know the value is always a boolean, but prevent all the other cases where we don't know and end up coercing.

**Comment by lifeart** [comment]
Don't know why GH showing conflict for `tslint.json`

**Comment by lifeart** [comment]
@krisselden, @pzuraq done.

File: packages/@glimmer/syntax/lib/builders.ts [hunk 3]
```diff
@@ -9,6 +9,10 @@ import { StringLiteral, BooleanLiteral, NumberLiteral } from './types/handlebars
 export type BuilderPath = string | AST.PathExpression;
 export type TagDescriptor = string | { name: string; selfClosing: boolean };
 
+function withFallback<T, U>(originalValue: T | null | undefined, fallbackValue: U): T | U {
```
Comment (pzuraq) [inline packages/@glimmer/syntax/lib/builders.ts]: We should consider changing this to nullish coalescing once TS 3.7 is released

**Pull request closed without merging**